liquidctl/driver/kraken_three_x.py

```diff
@@ -0,0 +1,327 @@
+    5: ([0x19, 0x00], [0x14, 0x00], [0x0f, 0x00], [0x07, 0x00], [0x04, 0x00]),
+    6: ([0x28, 0x00], [0x1e, 0x00], [0x14, 0x00], [0x0a, 0x00], [0x04, 0x00]),
+    7: ([0x32, 0x00], [0x28, 0x00], [0x1e, 0x00], [0x14, 0x00], [0x0a, 0x00]),
+    8: ([0x14, 0x00], [0x14, 0x00], [0x14, 0x00], [0x14, 0x00], [0x14, 0x00]),
+    9: ([0x00, 0x00], [0x00, 0x00], [0x00, 0x00], [0x00, 0x00], [0x00, 0x00]),
+    10:([0x37, 0x00], [0x28, 0x00], [0x19, 0x00], [0x0a, 0x00], [0x00, 0x00]),
+    11:([0x6e, 0x00], [0x53, 0x00], [0x39, 0x00], [0x2e, 0x00], [0x20, 0x00]),
+}
+_ANIMATION_SPEEDS = {
+    'slowest': 0x0,
+    'slower': 0x1,
+    'normal': 0x2,
+    'faster': 0x3,
+    'fastest': 0x4,
+}
+_ACCESSORY_NAMES = {
+    0x01: "HUE+ LED Strip",
+    0x02: "AER RGB 1",
+    0x04: "HUE 2 LED Strip 300 mm",
+    0x05: "HUE 2 LED Strip 250 mm",
+    0x06: "HUE 2 LED Strip 200 mm",
+    0x08: "HUE 2 Cable Comb",
+    0x0a: "HUE 2 Underglow 200 mm",
+    0x0b: "AER RGB 2 120 mm",
+    0x0c: "AER RGB 2 140 mm",
+    0x10: "Kraken X3 Pump Ring",
+    0x11: "Kraken X3 Pump Logo",
+}
+
+
+class KrakenThreeXDriver(UsbHidDriver):
+    """liquidctl driver for Kraken X3 devices from NZXT."""
+
+    SUPPORTED_DEVICES = [
+        (0x1e71, 0x2007, None, 'NZXT Kraken X3 Pump (X53, X63 or X73) (experimental)', {})
+    ]
+
+    def initialize(self, **kwargs):
+        """Initialize the device.
+
+        Reports the current firmware of the device.
+
+        Returns a list of (key, value, unit) tuples.
+        """
+        self.device.clear_enqueued_reports()
+        # request static infos
+        self._write([0x10, 0x01])  # firmware info
+        self._write([0x20, 0x03])  # lighting info
+        # initialize
+        self._write([0x70, 0x02, 0x01, 0xb8, 0x0b])
+        self._write([0x70, 0x01])
+        status = []
+
+        def parse_firm_info(msg):
+            fw = '{}.{}.{}'.format(msg[0x11], msg[0x12], msg[0x13])
+            status.append(('Firmware version', fw, ''))
+
+        def parse_led_info(msg):
+            """
+            FIXME: is is possible to attach other accessories to the pump?
+            currently not possible to address devices via led id e.g. "led1"
+            accessory_id: ? (LED 1 - ?) & 0x10 (LED 2 - ring) & 0x11 (LED 3 - logo)
+            """
+            num_light_channels = msg[14]  # the 15th byte (index 14) is # of light channels
+            accessories_per_channel = 6  # each lighting channel supports up to 6 accessories
+            light_accessory_index = 15  # offset in msg of info about first light accessory
+            for light_channel in range(num_light_channels):
+                for accessory_num in range(accessories_per_channel):
+                    accessory_id = msg[light_accessory_index]
+                    light_accessory_index += 1
+                    if accessory_id != 0:
+                        status.append(('LED {} accessory {}'.format(light_channel + 1, accessory_num + 1),
```

Comment:
After looking at the output with some accessories connected, I think the following output makes more sense:

```
$ liquidctl initialize
NZXT Kraken X (X53, X63 or X73) (experimental)
├── Firmware version                   1.8.0
├── LED accessory 1           HUE+ LED Strip
└── LED accessory 2           HUE+ LED Strip
```

 - remove (what used to be) the channel number
 - don't mention the pump or the ring, since they are not removable
 - but since we have this information, add an `assert` to check that both (pump and ring) have been reported as recognized

Reply:
It turns out that asserting would be annoying because the rest of the code only lists whatever accessories are there, which is significantly different from checking if two specific accessories are present (in specific channels too).

So instead I decided to simply log the additional accessories with level INFO (only shown for `--verbose` or `--debug`).

15a5cfe, then a0daa1a